code/datums/components/barber.dm

```diff
@@ -61,21 +60,67 @@ TYPEINFO(/datum/component/toggle_tool_use)
 
 TYPEINFO(/datum/component/barber)
 	initialization_args = list()
+	var/list/all_hairs = list()
+
+TYPEINFO_NEW(/datum/component/barber)
+	. = ..()
 
+	// just so we get a special icon sprite for no hair
+	src.all_hairs += list("None" = list("hair_id" = "none", "hair_icon" = "data:image/png;base64," + icon2base64(icon('icons/map-editing/landmarks.dmi', "x", SOUTH)), "hair_type" = /datum/customization_style/none))
+
+ABSTRACT_TYPE(/datum/component/barber)
 /datum/component/barber
+	var/datum/appearanceHolder/new_AH
+	var/datum/movable_preview/character/preview
+	var/mob/living/carbon/human/barbee
+	var/mob/barber
+	var/hair_portion = "bottom"
+
 /datum/component/barber/Initialize()
 	. = ..()
+
 	if(!istype(parent, /obj/item))
 		return COMPONENT_INCOMPATIBLE
 
+TYPEINFO(/datum/component/barber/haircut)
+	initialization_args = list()
+
+TYPEINFO_NEW(/datum/component/barber/haircut)
+	. = ..()
+
+
+	var/all_hair_types = concrete_typesof(/datum/customization_style/hair)
+	for (var/datum/customization_style/styles as anything in all_hair_types)
+		var/datum/customization_style/style = new styles // you have to initialize the datum before accessing it's values
+		var/hair_icon = "data:image/png;base64," + icon2base64(icon('icons/mob/human_hair.dmi', style.id, SOUTH, 1)) // yeah, sure, i'll keep it white. the user can preview the hair style anyway.
+		src.all_hairs += list(style.name = list("hair_id" = style.id, "hair_icon" = hair_icon, "hair_type" = style.type))
+
 /datum/component/barber/haircut
 /datum/component/barber/haircut/Initialize()
 	. = ..()
+	if (. == COMPONENT_INCOMPATIBLE)
+		return .
+
 	RegisterSignal(parent, COMSIG_ITEM_ATTACK_PRE, .proc/do_haircut)
 
+TYPEINFO(/datum/component/barber/shave)
+	initialization_args = list()
+
+TYPEINFO_NEW(/datum/component/barber/shave)
+	. = ..()
+
+	var/all_hair_types = concrete_typesof(/datum/customization_style/beard) + concrete_typesof(/datum/customization_style/moustache) + concrete_typesof(/datum/customization_style/sideburns) + concrete_typesof(/datum/customization_style/eyebrows)
```

Comment:
why not just have this as a var on the TYPEINFO and not repeat the icon init code?

Reply:
`concrete_typesof` is a proc, and you can't use procs when defining a var outside of a proc, which `TYPEINFO()` isn't. on an up-coming commit, i try to change a bit of the above